### PR TITLE
Implement the joint_fw_bw export path in AOTAutograd

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Summary:
This PR introduces an experimental training mode for dynamo.export().
It implements a export path for joint_fw_bw inside AOTAutograd. 

We also introduces a set of constrains to the exported program, such that the resulting joint graph would have the following signature.

        class JointGraphSignature:
            # joint graph inputs
            num_lifted_parameters: int
            num_lifted_buffers: int
            num_user_inputs: int

            # joint graph outputs
            num_input_mutations: int
            num_users_outputs: int
            num_gradient_parameters: int
            num_gradient_buffers: int
            num_gradient_user_inputs: int

A sample of the joint graph: https://www.internalfb.com/phabricator/paste/view/P629981166

Differential Revision: D43215635

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -2819,5 +2819,170 @@ def forward(*runtime_args):
...
+                out = mod(*args[params_len:], **kwargs)
+
+        # backward
+        assert out[0].requires_grad == True
+        assert out[0].numel() == 1, "Must be a scalar"
+
+        setup_stacktrace_preservation_hooks([out[0].grad_fn])
+
+        with fx_traceback.preserve_node_meta():
+            backward_out = torch.autograd.grad(out[0], args[:len(named_parameters)])
```
Comment (SherlockNoMad) [inline torch/_functorch/aot_autograd.py:2862]: Obviously, out[0] is a hack here. 

We have 3 options for UX of training mode export. 
1. Assuming out[0] is the loss to backprob with, so user must write the forward function to always return loss as the first return. 
2. Ask user to pass in an index to indicate which output is the loss value.
3. We can require that only one of the output can can require_grad=True. User would have to call .detach on the output that is not the loss. 

Ideas?

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -2819,5 +2819,170 @@ def forward(*runtime_args):
...
+
```
Comment (SherlockNoMad) [inline torch/_functorch/aot_autograd.py:2863]: `args[:len(named_parameters)]` is only back-propagating to parameters, and not to buffers and user inputs. 
Do we need to support computing gradients for user inputs? @suo 

I think it's needed as SparseArch output is fed as input to dense part as user input. Gradient needs to further backprob to SparseArch.

File: torch/_functorch/aot_autograd.py [hunk 3]
```diff
@@ -2819,5 +2819,170 @@ def forward(*runtime_args):
...
+            # joint graph inputs
+            num_lifted_parameters: int
+            num_lifted_buffers: int
+            num_user_inputs: int
+
+            # joint graph outputs
+            num_input_mutations: int
+            num_users_outputs: int
+            num_gradient_parameters: int
+            num_gradient_buffers: int
```
Comment (SherlockNoMad) [inline torch/_functorch/aot_autograd.py:2944]: we can drop the gradient for buffers, as buffers cannot have gradient.

File: torch/_functorch/aot_autograd.py [hunk 4]
```diff
@@ -2819,5 +2819,170 @@ def forward(*runtime_args):
...
+        )
+
+    cross_ref = CrossRefFakeMode() if config.debug_fake_cross_ref else nullcontext()
+    python_dispatcher_mode = (
+        enable_python_dispatcher() if shape_env is not None else nullcontext()
+    )
+
+    with torch.autograd.set_multithreading_enabled(
+        False
+    ), preserve_rng_state(), cross_ref, fake_mode, python_dispatcher_mode:
```
Comment (SherlockNoMad) [inline torch/_functorch/aot_autograd.py:2925]: Do I really need all these context? 
I don't think I need cross_ref. What about other?

**Review by ezyang — CHANGES_REQUESTED** [review]
I said it before and I'll say it again, please don't use argument position (e.g., counting the number of user outputs / gradient params / etc) to differentiate between params/buffers/user inputs. Force a more structured in/out representation.

**Comment by bdhirsh** [comment]
One thing that seems like it's missing: do you want to add extra asserts for the classes of model that export currently doesn't allow?

(1) Mutating an input that requires grad. Technically this is fine under `torch.no_grad()`, but I don't think we have an easy way to distinguish that case right now (but maybe this isn't a problem if you aren't exporting the optimizer step). This works today in aot autograd, but causes extra `.clone()`'s to show up in the graph

(2) Metadata mutations on inputs. You could probably in theory append `.as_strided_()` calls to the end of the graph if you detect input mutations, but that doesn't happen today (and another case where that isn't enough is if you see a `.resize_()` on an input).

(3) Mutating an input that aliases another input. The logic that checks for this today is in `merge_view_inputs`. Today, the second output from that function is an optional dict, and if the dict exists, then you know you hit that case (I think this could be refactored better though 🙁 )

File: torch/_dynamo/eval_frame.py [hunk 5]
```diff
@@ -519,7 +519,7 @@ def guard_export_print(guards):
 
 
 def export(
-    f, *args, aten_graph=False, decomposition_table=None, tracing_mode="real", **kwargs
+    f, *args, aten_graph=False, decomposition_table=None, tracing_mode="real", training_mode=False, **kwargs
```
Comment (voznesenskym) [inline torch/_dynamo/eval_frame.py:522]: Please no more notions and forks in export, let's start the unbundling yesterday?

File: torch/_functorch/aot_autograd.py [hunk 6]
```diff
@@ -2819,5 +2819,183 @@ def forward(*runtime_args):
...
+
+    named_parameters = dict(mod.named_parameters(remove_duplicate=False))
+    named_buffers = dict(mod.named_buffers(remove_duplicate=False))
+    params = {
+        **named_parameters,
+        **named_buffers
+    }
+    params_flat, params_spec = pytree.tree_flatten(params)
+    params_flat = tuple(params_flat)
+    params_len = len(params_flat)
```
Comment (voznesenskym) [inline torch/_functorch/aot_autograd.py:2838]: Dupe of L2571 in `aot_module_simplified` please pull into util

**Review by voznesenskym — CHANGES_REQUESTED** [review]
Paused reviewing - there's too much dupe code. If there's one thing we don't need, it's a bigger aot_autograd.py

If you need to have some upstream PRs to pull utils out of common places in aot_autograd, so be it!

Also, PR is untested.

**Comment by ezyang** [comment]
The resolution from Fri meeting was that s-curve will check in a private version of their code and @bdhirsh will do the refactor.
